test/main/modules.js

```diff
@@ -1,178 +1,386 @@
 /* global setTimeout */
 QUnit.module('QUnit.module', function () {
-  QUnit.module('before/beforeEach/afterEach/after', {
-    before: function () {
-      this.lastHook = 'module-before';
-    },
-    beforeEach: function (assert) {
-      assert.strictEqual(this.lastHook, 'module-before',
-        "Module's beforeEach runs after before");
-      this.lastHook = 'module-beforeEach';
-    },
-    afterEach: function (assert) {
-      assert.strictEqual(this.lastHook, 'test-block',
-        "Module's afterEach runs after current test block");
-      this.lastHook = 'module-afterEach';
-    },
-    after: function (assert) {
-      assert.strictEqual(this.lastHook, 'module-afterEach',
-        "Module's afterEach runs before after");
-      this.lastHook = 'module-after';
+  function flat (obj) {
+    var str = [];
+    for (var key in obj) {
+      str.push(key + '=' + obj[key]);
     }
-  });
+    return str.join(' ') || '(empty)';
+  }
+
+  var actual = {};
+  var expected = {
+    'parent with children': [
+      // parent > child > one
+      // parent > child > two
+      'parent-before: (empty)',
+      'child-before: before=P',
+      'parent-beforeEach: before=PC',
+      'child-beforeEach: before=PC beforeEach=P',
+      'child-test: before=PC beforeEach=PC',
+      'child-afterEach: before=PC beforeEach=PC tester=1',
+      'parent-afterEach: before=PC beforeEach=PC tester=1 afterEach=C',
+      'parent-beforeEach: before=PC',
+      'child-beforeEach: before=PC beforeEach=P',
+      'child-test: before=PC beforeEach=PC',
+      'child-afterEach: before=PC beforeEach=PC tester=2',
+      'parent-afterEach: before=PC beforeEach=PC tester=2 afterEach=C',
+      'child-after: before=PC beforeEach=PC tester=2 afterEach=CP',
+      'parent-after: before=PC beforeEach=PC tester=2 afterEach=CP after=C'
```

Comment:
with the changes in #1762 the `after` hooks will lose access to everything set in hooks other than `before` and `after` so I expect these will become
```
'child-after: before=PC',
'parent-after: before=P'
```

Perhaps it would be worth showing that multiple `after` hooks on the same level can see the previous env changes? ie

```
'child-after1: before=PC',
'child-after2: before=PC after=CA1',
'parent-after1: before=P',
'parent-after2: before=P after=PA1'
```

Reply:
Good idea! I'll add that.

Reply:
Yeah, "afterEach" should tear down state from beforeEach or tests. "after" would only see state that started at "begin". I do note that this differs from how e.g. PHPUnit works where a single `$this` by-reference persists throughout a single module (TestClase class).

But, I think it makes sense for QUnit because we already don't expose treat the state of other test cases with "after" callback either. All tests get their env garbage-collected of without after hooks, only the last one, which indeed looks wrong today as there should not be anything special about the properties the last test creates vs those from any previous test.

RE: Adding multiple hooks of the same. Good idea! I'll add that.